### ISSUE TITLE
Improve support for Input/OutputStream and non Marshallable objects

### ISSUE THREAD

**Issue by peter-lawrey** [OpenHFT/Chronicle-Wire#111]

**Comment by peter-lawrey** [comment]
This issue relates to https://stackoverflow.com/questions/53543494/how-to-read-data-serialized-with-chronicle-wire-from-inputstream/53543808#53543808

**Comment by Stephan972** [comment]
Hello Peter,

Firstly, thank you for quick support. I used the code posted on the [stackoverflow question](https://stackoverflow.com/questions/53543494/how-to-read-data-serialized-with-chronicle-wire-from-inputstream/53543808#53543808). It works fine.

However, some errors appear with the following wire types: CSV and READ_ANY. 

* CSV
net.openhft.chronicle.bytes.util.DecoratedBufferUnderflowException: readCheckOffset0 failed. Offset: -1 < start: 0
	at net.openhft.chronicle.bytes.AbstractBytes.readCheckOffset0(AbstractBytes.java:691)
	at net.openhft.chronicle.bytes.AbstractBytes.readCheckOffset(AbstractBytes.java:685)
	at net.openhft.chronicle.bytes.AbstractBytes.readByte(AbstractBytes.java:644)
	at net.openhft.chronicle.bytes.AbstractBytes.readUnsignedByte(AbstractBytes.java:315)
	at net.openhft.chronicle.wire.TextWire$TextValueIn.peekBack(TextWire.java:2230)
	at net.openhft.chronicle.wire.CSVWire$CSVValueIn.textTo0(CSVWire.java:204)
	at net.openhft.chronicle.wire.CSVWire$CSVValueIn.textTo0(CSVWire.java:140)
	at net.openhft.chronicle.wire.TextWire$TextValueIn.text(TextWire.java:2076)
	at net.openhft.chronicle.wire.CSVWire.<init>(CSVWire.java:41)
	at net.openhft.chronicle.wire.CSVWire.<init>(CSVWire.java:45)
	at net.openhft.chronicle.wire.WireType$10.apply(WireType.java:342)
	at net.openhft.chronicle.wire.WireType$10.apply(WireType.java:338)
	WireToOutputStream.<init>(first statement in the constructor)

* READ_ANY
java.lang.NullPointerException
	at net.openhft.chronicle.wire.AbstractAnyWire.getValueOut(AbstractAnyWire.java:170)
	at <...>testVisSocket(WireToOutputStreamTest.java:26)

**Comment by Stephan972** [comment]
More over, taking the [testVisSocket](https://github.com/OpenHFT/Chronicle-Wire/commit/bd93763240adec7cd867b8932c144f6be3f67f6f#diff-2b1bccbbf4dec52ff385a66e8016a315) and adding two java.sql.Timestamp fields* to AnObject, all wire types fail except TEXT and JSON.

_* One field is set to new Timestamp(1234567890) and the other is always null._ 

* BINARY, BINARY_LIGHT, COMPRESSED_BINARY (same stacktrace error)

java.lang.NumberFormatException: For input string: "# position: 88, header: 0\r\n#  has a 4 byte size prefix, 2946 > 11 len is 2946"
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.lang.Double.parseDouble(Double.java:538)
	at net.openhft.chronicle.wire.BinaryWire$BinaryValueIn.readTextAsLong(BinaryWire.java:3341)
	at net.openhft.chronicle.wire.BinaryWire$BinaryValueIn.int64(BinaryWire.java:3458)
	at net.openhft.chronicle.wire.Wires.objectDate(Wires.java:447)
	at net.openhft.chronicle.wire.Wires.object0(Wires.java:488)
	at net.openhft.chronicle.wire.BinaryWire$BinaryValueIn.object(BinaryWire.java:2088)
	at net.openhft.chronicle.wire.BinaryWire$BinaryValueIn.objectWithInferredType(BinaryWire.java:3573)
	at net.openhft.chronicle.wire.ValueIn.object(ValueIn.java:580)
	at net.openhft.chronicle.wire.WireMarshaller$ObjectFieldAccess.setValue(WireMarshaller.java:503)
	at net.openhft.chronicle.wire.WireMarshaller$FieldAccess.readValue(WireMarshaller.java:400)
	at net.openhft.chronicle.wire.WireMarshaller.readMarshallable(WireMarshaller.java:171)
	at net.openhft.chronicle.wire.Wires.readMarshallable(Wires.java:315)
	at <...>testVisSocket(WireToOutputStreamTest.java:34)

* FIELDLESS_BINARY

java.lang.NumberFormatException: For input string: "# position: 66, header: 0\r\n#  has a 4 byte size prefix, 1410 > 5 len is 1410"
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.lang.Double.parseDouble(Double.java:538)
	at net.openhft.chronicle.wire.BinaryWire$BinaryValueIn.readTextAsLong(BinaryWire.java:3341)
	at net.openhft.chronicle.wire.BinaryWire$BinaryValueIn.int64(BinaryWire.java:3458)
	at net.openhft.chronicle.wire.Wires.objectDate(Wires.java:447)
	at net.openhft.chronicle.wire.Wires.object0(Wires.java:488)
	at net.openhft.chronicle.wire.BinaryWire$BinaryValueIn.object(BinaryWire.java:2088)
	at net.openhft.chronicle.wire.BinaryWire$BinaryValueIn.objectWithInferredType(BinaryWire.java:3573)
	at net.openhft.chronicle.wire.ValueIn.object(ValueIn.java:580)
	at net.openhft.chronicle.wire.WireMarshaller$ObjectFieldAccess.setValue(WireMarshaller.java:503)
	at net.openhft.chronicle.wire.WireMarshaller$FieldAccess.readValue(WireMarshaller.java:400)
	at net.openhft.chronicle.wire.WireMarshaller.readMarshallable(WireMarshaller.java:171)
	at net.openhft.chronicle.wire.Wires.readMarshallable(Wires.java:315)
	at <...>(WireToOutputStreamTest.java:34)

* RAW

java.lang.UnsupportedOperationException
	at net.openhft.chronicle.wire.RawWire$RawValueIn.objectWithInferredType(RawWire.java:1141)
	at net.openhft.chronicle.wire.ValueIn.object(ValueIn.java:580)
	at net.openhft.chronicle.wire.WireMarshaller$ObjectFieldAccess.setValue(WireMarshaller.java:503)
	at net.openhft.chronicle.wire.WireMarshaller$FieldAccess.readValue(WireMarshaller.java:400)
	at net.openhft.chronicle.wire.WireMarshaller.readMarshallable(WireMarshaller.java:171)
	at net.openhft.chronicle.wire.Wires.readMarshallable(Wires.java:315)
	at <...>testVisSocket(WireToOutputStreamTest.java:19)

* READ_ANY

java.lang.NullPointerException
	at net.openhft.chronicle.wire.AbstractAnyWire.getValueOut(AbstractAnyWire.java:170)
	at <...>testVisSocket(WireToOutputStreamTest.java:26)

How can we solve those issues (here and previous post) ?

**Comment by peter-lawrey** [comment]
I will have a look next week.

**Comment by MartyIX** [comment]
As for CSV case (no header is provided):

1) The exception in `peekBack` method is caused by https://github.com/OpenHFT/Chronicle-Wire/blob/master/src/main/java/net/openhft/chronicle/wire/TextWire.java#L2236 where the inequality condition should simply be `>` instead of  `>=`. I believe this is a bug in any case.

2) After fixing 1), one will get an infinite cycle here https://github.com/OpenHFT/Chronicle-Wire/blob/master/src/main/java/net/openhft/chronicle/wire/CSVWire.java#L41 as there are no data in `bytes` parameter. I tried a hotfix:

```java
    public CSVWire(@NotNull Bytes bytes, boolean use8bit) {
        super(bytes, use8bit);
        while (lineStart == 0) {
            String text = valueIn.text();

            // ADDED
            if (text == null || text.isEmpty()) {
                break;
            }

            header.add(text);
        }
    }
```

3) Then WireMarshaller tries to set fields of `AnObject` from string `value: 12345\ntext: Hello\n`. TextWire::int64() is being called. The problem is that `value: ` is not being skipped. And that's because https://github.com/OpenHFT/Chronicle-Wire/blob/master/src/main/java/net/openhft/chronicle/wire/CSVWire.java#L117 ignores the key it gets.

I'm not sure whether CSVWire is supposed to support read/write marshalled objects though.

**Comment by hft-team-city** [comment]
Released in [Chronicle-Wire-2.17.12](https://github.com/OpenHFT/Chronicle-Wire/releases/chronicle-wire-2.17.12), [BOM-2.17.196](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.196)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Wire-2.17.12](https://github.com/OpenHFT/Chronicle-Wire/releases/chronicle-wire-2.17.12), [BOM-2.17.225](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.225)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Wire-2.17.12](https://github.com/OpenHFT/Chronicle-Wire/releases/chronicle-wire-2.17.12), [BOM-2.17.241](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.241)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Wire-2.20.101](https://github.com/OpenHFT/Chronicle-Wire/releases/chronicle-wire-2.20.101), [BOM-2.20.134](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.20.134)
